File: src/ru/serjik/wallpaper/BaseSettingsActivity.java
package ru.serjik.wallpaper;

import android.app.Activity;
import android.app.WallpaperInfo;
import android.app.WallpaperManager;
import android.content.ActivityNotFoundException;
import android.content.ComponentName;
import android.content.Intent;
import android.view.View;
import android.widget.Toast;

public abstract class BaseSettingsActivity extends Activity {
    private int savedSystemUiVisibility = 0;

    private View.OnClickListener rootClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            view.setOnClickListener(null);
            view.setClickable(false);
            showUI();
        }
    };

    protected void showUI() {
        getWindow().getDecorView().setSystemUiVisibility(this.savedSystemUiVisibility);
        getWindow().getDecorView().findViewWithTag("ui").setVisibility(View.VISIBLE);
    }

    private void sendDropContextCommand() {
        Intent intent = new Intent(this, getWallpaperServiceClass());
        intent.putExtra("cmd", "dropContext");
        startService(intent);
    }

    private void setupHideButton() {
        View hideButton = getWindow().getDecorView().findViewWithTag("button_hide_ui");
        if (hideButton != null) {
            hideButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    hideUI();
                }
            });
        }
    }

    private void setupSetWallpaperButton() {
        View setWallpaperButton = getWindow().getDecorView().findViewWithTag("button_set_wallpaper");
        if (setWallpaperButton != null) {
            setWallpaperButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (isTaskRoot()) {
                        setAsWallpaper();
                    }
                    finish();
                }
            });
            setWallpaperButton.setVisibility(isWallpaperAlreadySet() ? View.GONE : View.VISIBLE);
        }
    }

    private boolean isWallpaperAlreadySet() {
        WallpaperManager wm = WallpaperManager.getInstance(this);
        WallpaperInfo info = wm.getWallpaperInfo();
        if (info != null) {
            if (getClass().getCanonicalName().equals(info.getSettingsActivity())) {
                wm.forgetLoadedWallpaper();
                return true;
            }
        }
        return false;
    }

    protected void setAsWallpaper() {
        try {
            ComponentName component = new ComponentName(getWallpaperServiceClass().getPackage().getName(), getWallpaperServiceClass().getCanonicalName());
            Intent intent = new Intent("android.service.wallpaper.CHANGE_LIVE_WALLPAPER");
            intent.putExtra("android.service.wallpaper.extra.LIVE_WALLPAPER_COMPONENT", component);
            startActivity(intent);
        } catch (ActivityNotFoundException e) {
            try {
                startActivity(new Intent("android.service.wallpaper.LIVE_WALLPAPER_CHOOSER"));
            } catch (ActivityNotFoundException e2) {
                try {
                    Intent intent = new Intent();
                    intent.setAction("com.bn.nook.CHANGE_WALLPAPER");
                    startActivity(intent);
                } catch (ActivityNotFoundException e3) {
                    Toast.makeText(getBaseContext(), "something goes wrong", Toast.LENGTH_LONG).show();
                }
            }
        }
    }

    protected abstract Class<?> getWallpaperServiceClass();

    protected void hideUI() {
        getWindow().getDecorView().findViewWithTag("root").setClickable(true);
        getWindow().getDecorView().findViewWithTag("root").setOnClickListener(this.rootClickListener);
        getWindow().getDecorView().findViewWithTag("ui").setVisibility(View.GONE);
        this.savedSystemUiVisibility = getWindow().getDecorView().getSystemUiVisibility();
        getWindow().getDecorView().setSystemUiVisibility(3847);
    }

    @Override
    protected void onPause() {
        super.onPause();
        sendDropContextCommand();
    }

    @Override
    protected void onResume() {
        super.onResume();
        setupSetWallpaperButton();
        setupHideButton();
    }
}

File: app/src/main/java/ru/serjik/wallpaper/BaseSettingsActivity.kt
package ru.serjik.wallpaper

import android.app.Activity
import android.app.WallpaperManager
import android.content.ActivityNotFoundException
import android.content.ComponentName
import android.content.Intent
import android.os.Build
import android.view.View
import android.view.WindowInsets
import android.view.WindowInsetsController
import android.widget.Toast

abstract class BaseSettingsActivity : Activity() {

    private val rootClickListener = View.OnClickListener { view ->
        view.setOnClickListener(null)
        view.isClickable = false
        showUI()
    }

    protected fun showUI() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.insetsController?.show(WindowInsets.Type.systemBars())
        } else {
            @Suppress("DEPRECATION")
            window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_VISIBLE
        }
        window.decorView.findViewWithTag<View>("ui").visibility = View.VISIBLE
    }

    /**
     * Notifies the wallpaper service to reload by writing a signal to SharedPreferences.
     * The service listens for changes via OnSharedPreferenceChangeListener.
     */
    private fun notifyServiceReload() {
        val prefs = getSharedPreferences("application_store", MODE_PRIVATE)
        prefs.edit()
            .putLong("reload_signal", System.currentTimeMillis())
            .apply()
    }

    private fun setupHideButton() {
        window.decorView.findViewWithTag<View>("button_hide_ui")?.setOnClickListener {
            hideUI()
        }
    }

    private fun setupSetWallpaperButton() {
        window.decorView.findViewWithTag<View>("button_set_wallpaper")?.apply {
            setOnClickListener {
                if (isTaskRoot) {
                    setAsWallpaper()
                }
                finish()
            }
            visibility = if (isWallpaperAlreadySet()) View.GONE else View.VISIBLE
        }
    }

    private fun isWallpaperAlreadySet(): Boolean {
        val wm = WallpaperManager.getInstance(this)
        val info = wm.wallpaperInfo ?: return false
        if (this::class.java.canonicalName == info.settingsActivity) {
            wm.forgetLoadedWallpaper()
            return true
        }
        return false
    }

    protected fun setAsWallpaper() {
        try {
            val component = ComponentName(
                wallpaperServiceClass.`package`!!.name,
                wallpaperServiceClass.canonicalName!!
            )
            val intent = Intent("android.service.wallpaper.CHANGE_LIVE_WALLPAPER").apply {
                putExtra("android.service.wallpaper.extra.LIVE_WALLPAPER_COMPONENT", component)
            }
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            try {
                startActivity(Intent("android.service.wallpaper.LIVE_WALLPAPER_CHOOSER"))
            } catch (e2: ActivityNotFoundException) {
                try {
                    startActivity(Intent().apply {
                        action = "com.bn.nook.CHANGE_WALLPAPER"
                    })
                } catch (e3: ActivityNotFoundException) {
                    Toast.makeText(baseContext, "something goes wrong", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    protected abstract val wallpaperServiceClass: Class<*>

    protected fun hideUI() {
        window.decorView.findViewWithTag<View>("root").apply {
            isClickable = true
            setOnClickListener(rootClickListener)
        }
        window.decorView.findViewWithTag<View>("ui").visibility = View.GONE
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.insetsController?.apply {
                hide(WindowInsets.Type.systemBars())
                systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            }
        } else {
            @Suppress("DEPRECATION")
            window.decorView.systemUiVisibility = (
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                )
        }
    }

    override fun onPause() {
        super.onPause()
        notifyServiceReload()
    }

    override fun onResume() {
        super.onResume()
        setupSetWallpaperButton()
        setupHideButton()
    }
}
